### PR TITLE
[crypto] Add a key manager driver and basic tests.

### PR DESCRIPTION
**Pull request by jadephilipoom**
- Adds a cryptolib driver for the key manager.
    - The cryptolib never needs to advance the keymgr state, so the driver only has the capability to generate/clear keys
- Adds a test for the basic functionality of the driver
    - I'll merge follow-up PRs with functional tests for each sideloaded block that actually use and clear the keys, but since these tests will involve specific details of each block it makes sense to me for them to be separate and live under `sw/device/tests/crypto` instead of with the driver.

### CODE REVIEW

File: sw/device/lib/crypto/drivers/keymgr.c [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+  // Should be unreachable.
+  HARDENED_TRAP();
+  return OTCRYPTO_FATAL_ERR;
+}
+
+/**
+ * Set the control register of the key manager.
+ *
+ * The CDI select bit is always set to false for this driver.
```
Comment (moidx) [inline sw/device/lib/crypto/drivers/keymgr.c:126]: nit: for documentation purposes, we should list this as SEALING_CDI. I assume that this change does not include support for attestation flows.

File: sw/device/lib/crypto/drivers/keymgr.c [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+status_t keymgr_generate_key_sw(keymgr_diversification_t diversification,
+                                keymgr_output_t *key) {
+  HARDENED_TRY(keymgr_is_idle());
+
+  // Set the control register to generate a software-visible key.
+  WRITE_CTRL(NONE, GENERATE_SW);
+
+  // Check that the version does not exceed the max owner key version.
+  uint32_t max_version =
+      abs_mmio_read32(kBaseAddr + KEYMGR_MAX_OWNER_KEY_VER_SHADOWED_REG_OFFSET);
```
Comment (moidx) [inline sw/device/lib/crypto/drivers/keymgr.c:154]: This assumes that the library will only be used when the `keymgr` has reached the `OWNER` stage. The `ROM_EXT` may have some use cases for generating keys, although it can probably limit use to sideloaded keys.

I think if you want to keep this code as is, it probably makes sense to check the current state of the `keymgr`. Otherwise I would recommend to use the current internal state to select the appropriate `max ver` register to compare against.

Side note: the hardware is expected to return an error if the diversification version is greater than the current `max_version`. This error may manifest itself as `INVALID_OP` with `DEBUG == INVALID_KEY_VERSION`

File: sw/device/lib/crypto/drivers/keymgr.c [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+  // Set SIDELOAD_CLEAR to begin continuously clearing the requested slot.
+  abs_mmio_write32(
+      kBaseAddr + KEYMGR_SIDELOAD_CLEAR_REG_OFFSET,
+      bitfield_field32_write(0, KEYMGR_SIDELOAD_CLEAR_VAL_FIELD, slot));
+
+  // Spin for 100 microseconds.
+  // TODO: this value seems to work for tests, but it would be good to run a
+  // more principled analysis.
+  busy_spin_micros(100);
```
Comment (moidx) [inline sw/device/lib/crypto/drivers/keymgr.c:236]: nit: For hardening, confirm that the expected value was written to `KEYMGR_SIDELOAD_CLEAR_REG_OFFSET`. I suspect that the action of reading back the register will give us enough time to perform the clear operation inside `keymgr`, but don't think it hurts to have the delay.

**Review by moidx — APPROVED** [review]

File: sw/device/lib/crypto/drivers/keymgr.c (hunk 2 above)
Comment (jadephilipoom) [reply sw/device/lib/crypto/drivers/keymgr.c:154]: My understanding was that the cryptolib drivers were exclusively for the OS-visible crypto library, and that `ROM_EXT` would use the drivers under `silicon_creator`. Or is it the case that the cryptolib drivers are for `ROM_EXT` and cryptolib, and the `silicon_creator` drivers are for `ROM` only? I suppose that makes sense from a binary separation standpoint.

As for why I checked the result in software here even though it would result in a hardware error: this is mostly just to help the caller by returning `OTCRYPTO_BAD_ARGS` instead of `OTCRYPTO_RECOV_ERR`, since I thought that was a more appropriate message for this case. Not absolutely necessary to do this, though, especially if it would complicate the code.

Comment (jadephilipoom) [reply sw/device/lib/crypto/drivers/keymgr.c:154]: Maybe for now I'll just remove the check; the caller will get an error anyway and this way we don't introduce the possibility of an error in a case that shouldn't actually cause one.

Comment (moidx) [reply sw/device/lib/crypto/drivers/keymgr.c]: I think it is ok to keep, but it would then make sense to check that the `keymgr` is in the expected state. 

I think in general it makes sense for `ROM_EXT` to use cryptolib, but it would be interesting to figure out first how we are planning to provide crypto functionality that overlaps both the Silicon Creator and Silicon Owner (e.g. attestation signatures). Other than attestation, I foresee the need for sideloading keys into KMAC for integrity calculations in `ROM_EXT`; and, entropy complex initialization. I am not sure if I am missing anything else.

File: sw/device/tests/keymgr_key_derivation_test.c [hunk 4]
```diff
@@ -61,6 +61,32 @@ bool test_main(void) {
...
+  sideload_params.dest = kDifKeymgrVersionedKeyDestAes;
+  CHECK_STATUS_OK(
+      keymgr_testutils_generate_versioned_key(&keymgr, sideload_params));
+  LOG_INFO("Keymgr generated HW output for Aes at OwnerRootKey State");
+
+  sideload_params.dest = kDifKeymgrVersionedKeyDestOtbn;
+  CHECK_STATUS_OK(
+      keymgr_testutils_generate_versioned_key(&keymgr, sideload_params));
+  LOG_INFO("Keymgr generated HW output for Otbn at OwnerRootKey State");
+
```
Comment (ballifatih) [inline sw/device/tests/keymgr_key_derivation_test.c:89]: Are these changes necessary on this top-level test? It seems to me like since we are not comparing the key outputs, we are not actually testing the correct generation of values. 

I wonder if its DV component (`chip_sw_keymgr_key_derivation_vseq.sv` and few other sequences derived from this) also needs to be updated so that we can confirm these keys are generated correctly.

**Review by ballifatih — APPROVED** [review]
LGTM, with a comment:

I am not sure whether top-level test changes might cause some tests to fail at nightly, as some printed messages from C test is used to synchronize on the DV side. I had a quick look and I am guessing those tests will probably not fail. However, no actual key comparison is made for the versioned keys.

File: sw/device/tests/keymgr_key_derivation_test.c (hunk 4 above)
Comment (jadephilipoom) [reply sw/device/tests/keymgr_key_derivation_test.c:89]: I suppose it's not really necessary; I just figured it made sense to test the `OwnerRootKey` state as well since it wasn't previously tested. Ideally the DV component would check the expected values, you're right; maybe I should remove it from this change to avoid giving the impression we're testing something we're really not testing.

**Comment by jadephilipoom** [comment]
Removed the top-level test changes and addressed Fatih's comments; I'll merge this now.
